File: streaming/python/executiongraph.py
import ray.streaming.generated.remote_call_pb2 as remote_call_pb


class ExecutionGraph:
    def __init__(self, graph_pb: remote_call_pb.ExecutionGraph):
        self._graph_pb = graph_pb

    def build_time(self):
        return self._graph_pb.build_time()

    def execution_nodes(self):
        return self._graph_pb.execution_nodes()

    def get_execution_task_by_task_id(self, task_id):
        for execution_node in self._graph_pb.execution_nodes:
            for task in execution_node.execution_tasks:
                if task.task_id == task_id:
                    return task
        raise Exception("Task %s does not exist!".format(task_id))

    def get_execution_node_by_task_id(self, task_id):
        for execution_node in self._graph_pb.execution_nodes:
            for task in execution_node.execution_tasks:
                if task.task_id == task_id:
                    return execution_node
        raise Exception("Task %s does not exist!".format(task_id))

    def get_task_id2_worker_by_node_id(self, node_id):
        for execution_node in self._graph_pb.execution_nodes:
            if execution_node.node_id == node_id:
                task_id2_worker = {}
                for task in execution_node.execution_tasks:
                    worker_actor_id_bytes = task.worker_actor_id
                    task_id2_worker[task.task_id] = worker_actor_id_bytes
                return task_id2_worker
        raise Exception("Node %s does not exist!".format(node_id))

File: streaming/java/streaming-runtime/src/main/java/org/ray/streaming/runtime/core/graph/ExecutionGraph.java
package org.ray.streaming.runtime.core.graph;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import org.ray.api.RayActor;

/**
 * Physical execution graph.
 *
 * <p>Notice: Temporary implementation for now to keep functional. This will be changed to
 * {@link org.ray.streaming.runtime.core.graph.executiongraph.ExecutionGraph} later when
 * new stream task implementation is ready.
 */
public class ExecutionGraph implements Serializable {
  private long buildTime;
  private List<ExecutionNode> executionNodeList;
  private List<RayActor> sourceWorkers = new ArrayList<>();
  private List<RayActor> sinkWorkers = new ArrayList<>();

  public ExecutionGraph(List<ExecutionNode> executionNodes) {
    this.executionNodeList = executionNodes;
    for (ExecutionNode executionNode : executionNodeList) {
      if (executionNode.getNodeType() == ExecutionNode.NodeType.SOURCE) {
        List<RayActor> actors = executionNode.getExecutionTasks().stream()
            .map(ExecutionTask::getWorker).collect(Collectors.toList());
        sourceWorkers.addAll(actors);
      }
      if (executionNode.getNodeType() == ExecutionNode.NodeType.SINK) {
        List<RayActor> actors = executionNode.getExecutionTasks().stream()
            .map(ExecutionTask::getWorker).collect(Collectors.toList());
        sinkWorkers.addAll(actors);
      }
    }
    buildTime = System.currentTimeMillis();
  }

  public List<RayActor> getSourceWorkers() {
    return sourceWorkers;
  }

  public List<RayActor> getSinkWorkers() {
    return sinkWorkers;
  }

  public List<ExecutionNode> getExecutionNodeList() {
    return executionNodeList;
  }

  public ExecutionTask getExecutionTaskByTaskId(int taskId) {
    for (ExecutionNode executionNode : executionNodeList) {
      for (ExecutionTask executionTask : executionNode.getExecutionTasks()) {
        if (executionTask.getTaskId() == taskId) {
          return executionTask;
        }
      }
    }
    throw new RuntimeException("Task " + taskId + " does not exist!");
  }

  public ExecutionNode getExecutionNodeByNodeId(int nodeId) {
    for (ExecutionNode executionNode : executionNodeList) {
      if (executionNode.getNodeId() == nodeId) {
        return executionNode;
      }
    }
    throw new RuntimeException("Node " + nodeId + " does not exist!");
  }

  public ExecutionNode getExecutionNodeByTaskId(int taskId) {
    for (ExecutionNode executionNode : executionNodeList) {
      for (ExecutionTask executionTask : executionNode.getExecutionTasks()) {
        if (executionTask.getTaskId() == taskId) {
          return executionNode;
        }
      }
    }
    throw new RuntimeException("Task " + taskId + " does not exist!");
  }

  public Map<Integer, RayActor> getTaskId2WorkerByNodeId(int nodeId) {
    for (ExecutionNode executionNode : executionNodeList) {
      if (executionNode.getNodeId() == nodeId) {
        Map<Integer, RayActor> taskId2Worker = new HashMap<>();
        for (ExecutionTask executionTask : executionNode.getExecutionTasks()) {
          taskId2Worker.put(executionTask.getTaskId(), executionTask.getWorker());
        }
        return taskId2Worker;
      }
    }
    throw new RuntimeException("Node " + nodeId + " does not exist!");
  }

  public long getBuildTime() {
    return buildTime;
  }
}
